src/opengl.c

```diff
@@ -889,6 +1237,303 @@ bool glx_blur_dst(session_t *ps, int dx, int dy, int width, int height, float z,
+	GLuint tex_scr = pbc->textures[0];
+
+	pbc->width = mwidth;
+	pbc->height = mheight;
+
+	if (!tex_scr) {
+		log_error("Failed to allocate texture.");
+		goto glx_round_corners_dst_end;
+	}
+
+	// Read destination pixels into a texture
+	glEnable(tex_tgt);
+	glBindTexture(tex_tgt, tex_scr);
+	glx_copy_region_to_tex(ps, tex_tgt, mdx, mdy, mdx, mdy, mwidth, mheight);
+
+	// Texture scaling factor
+	GLfloat texfac_x = 1.0f, texfac_y = 1.0f;
+	if (tex_tgt == GL_TEXTURE_2D) {
+		texfac_x /= (GLfloat)mwidth;
+		texfac_y /= (GLfloat)mheight;
+	}
+
+	// Paint it back
+	{
+		glDisable(GL_STENCIL_TEST);
+		glDisable(GL_SCISSOR_TEST);
+	}
+
+	{
+		const glx_round_pass_t *ppass = &ps->psglx->round_passes[shader_idx];
+		assert(ppass->prog);
+
+		assert(tex_scr);
+		glBindTexture(tex_tgt, tex_scr);
+
+		glBindFramebuffer(GL_FRAMEBUFFER, 0);
+		glDrawBuffer(GL_BACK);
+		if (have_scissors)
+			glEnable(GL_SCISSOR_TEST);
+		if (have_stencil)
+			glEnable(GL_STENCIL_TEST);
+
+		// Our shader generates a transparent mid section
+		// with opaque corners copied from the background texture
+		// We must use blending to get the window pixels to appear
+		glEnable(GL_BLEND);
+		glBlendFunc(GL_ONE, GL_ONE_MINUS_SRC_ALPHA);
+
+		glTexEnvi(GL_TEXTURE_ENV, GL_TEXTURE_ENV_MODE, GL_REPLACE);
+
+		glUseProgram(ppass->prog);
+
+		// If caller specified a texture use it as source
+		if (ptex) {
+			log_debug("ptex: %p wh(%d %d) %d %d", ptex, ptex->width,
+			          ptex->height, ptex->target, ptex->texture);
+			bind_sampler_to_unit_with_texture(ppass->prog, "tex_scr", 0,
+			                                  ptex->target, ptex->texture);
+		} else {
+			bind_sampler_to_unit_with_texture(ppass->prog, "tex_scr", 0,
+			                                  tex_tgt, tex_scr);
+		}
+
+		if (ppass->unifm_radius >= 0)
+			glUniform1f(ppass->unifm_radius, cr);
+		if (ppass->unifm_texcoord >= 0)
+			glUniform2f(ppass->unifm_texcoord, (float)dx, (float)dy);
+		if (ppass->unifm_texsize >= 0)
+			glUniform2f(ppass->unifm_texsize, (float)mwidth, (float)mheight);
+		if (ppass->unifm_borderw >= 0)
+			glUniform1f(ppass->unifm_borderw, w->g.border_width);
+		if (ppass->unifm_is_focused >= 0)
+			glUniform1i(ppass->unifm_is_focused, w->focused);
+		if (ppass->unifm_resolution >= 0)
+			glUniform2f(ppass->unifm_resolution, (float)ps->root_width,
+			            (float)ps->root_height);
+
+		// Painting
+		{
+			P_PAINTREG_START(crect) {
+				// XXX explain these variables
+				auto rx = (GLfloat)(crect.x1 - dx);
+				auto ry = (GLfloat)(crect.y1 - dy);
+				auto rxe = rx + (GLfloat)(crect.x2 - crect.x1);
+				auto rye = ry + (GLfloat)(crect.y2 - crect.y1);
+				// Rectangle textures have [0-w] [0-h] while 2D texture
+				// has [0-1] [0-1] Thanks to amonakov for pointing out!
+				if (GL_TEXTURE_2D == tex_tgt) {
+					rx = rx / (GLfloat)width;
+					ry = ry / (GLfloat)height;
+					rxe = rxe / (GLfloat)width;
+					rye = rye / (GLfloat)height;
+				}
```

Comment:
IIRC, these are texture-local coordinates, either normalized `[0-1]` or not `[0-w]`/`[0-h]` depending on the texture sampler, for the source texture.